nativelib/src/main/resources/gc/commix/Allocator.c

```diff
@@ -0,0 +1,427 @@
+}
+
+INLINE
+word_t *Allocator_lazySweep(Heap *heap, uint32_t size) {
+    word_t *object = NULL;
+    Stats_DefineOrNothing(stats, heap->stats);
+    Stats_RecordTime(stats, start_ns);
+    // mark as active
+    heap->lazySweep.lastActivity = BlockRange_Pack(1, heap->sweep.cursor);
+    while (object == NULL && heap->sweep.cursor < heap->sweep.limit) {
+        Sweeper_Sweep(heap, heap->stats, &heap->lazySweep.cursorDone,
+                      LAZY_SWEEP_MIN_BATCH);
+        object = Allocator_tryAlloc(&allocator, size);
+    }
+    // mark as inactive
+    heap->lazySweep.lastActivity = BlockRange_Pack(0, heap->sweep.cursor);
+    while (object == NULL && !Sweeper_IsSweepDone(heap)) {
+        object = Allocator_tryAlloc(&allocator, size);
+        if (object == NULL) {
+            sched_yield();
+        }
+    }
+    Stats_RecordTime(stats, end_ns);
+    Stats_RecordEvent(stats, event_sweep, start_ns, end_ns);
+    return object;
+}
+
+NOINLINE word_t *Allocator_allocSlow(Heap *heap, uint32_t size) {
+    word_t *object = Allocator_tryAlloc(&allocator, size);
+
+    if (object != NULL) {
+    done:
+        assert(Heap_IsWordInHeap(heap, object));
+        assert(object != NULL);
+        ObjectMeta *objectMeta = Bytemap_Get(allocator.bytemap, object);
+#ifdef DEBUG_ASSERT
+        ObjectMeta_AssertIsValidAllocation(objectMeta, size);
+#endif
+        ObjectMeta_SetAllocated(objectMeta);
+        return object;
+    }
+
+    if (!Sweeper_IsSweepDone(heap)) {
+        object = Allocator_lazySweep(heap, size);
+
+        if (object != NULL)
+            goto done;
+    }
+
+    Heap_Collect(heap);
+    object = Allocator_tryAlloc(&allocator, size);
+
+    if (object != NULL)
+        goto done;
+
+    if (!Sweeper_IsSweepDone(heap)) {
+        object = Allocator_lazySweep(heap, size);
+
+        if (object != NULL)
+            goto done;
+    }
+
+    // A small object can always fit in a single free block
+    // because it is no larger than 8K while the block is 32K.
+    Heap_Grow(heap, 1);
+    object = Allocator_tryAlloc(&allocator, size);
+
+    goto done;
+}
+
+INLINE word_t *Allocator_Alloc(Heap *heap, uint32_t size) {
+    assert(size % ALLOCATION_ALIGNMENT == 0);
+    assert(size < MIN_BLOCK_SIZE);
+
+    word_t *start = allocator.cursor;
+    word_t *end = (word_t *)((uint8_t *)start + size);
+
+    // Checks if the end of the block overlaps with the limit
+    if (end >= allocator.limit) {
+        return Allocator_allocSlow(heap, size);
+    }
+
+    allocator.cursor = end;
+
+    memset(start, 0, size);
+
+    word_t *object = start;
+    ObjectMeta *objectMeta = Bytemap_Get(allocator.bytemap, object);
+#ifdef DEBUG_ASSERT
+    ObjectMeta_AssertIsValidAllocation(objectMeta, size);
+#endif
+    ObjectMeta_SetAllocated(objectMeta);
+
+    __builtin_prefetch(object + 36, 0, 3);
+
+    assert(Heap_IsWordInHeap(heap, object));
+    return object;
+}
+
+uint32_t Allocator_Sweep(Allocator *allocator, BlockMeta *blockMeta,
```

Comment:
1. Why do we have sweeping logic in allocator?
2. Is this function used?

Reply:
Both `Allocator_Sweep` and `LargeAllocator_Sweep` are called from `Sweeper_Sweep`.
1. They take their corresponding allocator as the first argument.
2. All the logic that can change the state of small object specific metadata is in `Allocator` and for large objects it is `LargeAllocator`. It is convenient and I see no benefit of  splitting it.
3. This saves LargeAllocator from exposing the chunk and its freelist manipulation logic which is part of its implementation.
4. The obvious alternative is to put them both in `Sweeper` and call them something like `Sweeper_SweepSmall` and `Sweeper_SweepLarge` which I do not like. For the same reason I moved `Allocator_lazySweep` and `LargeAllocator_lazySweep` to these modules.

Reply:
Having sweeping logic spread over multiple modules makes it harder to see the whole picture. This needs to be refactored better for clarity.

Reply:
You are right. After all we do not split the conservative marking logic in `Marker` either. I moved both functions to `Sweeper`. And they are private - not exposed in the `*.h` file.